src/transformers/modeling_bert.py

```diff
@@ -953,26 +1065,12 @@ def forward(
 
         outputs = (prediction_scores,) + outputs[2:]  # Add hidden states and attention if they are here
 
-        # Although this may seem awkward, BertForMaskedLM supports two scenarios:
-        # 1. If a tensor that contains the indices of masked labels is provided,
-        #    the cross-entropy is the MLM cross-entropy that measures the likelihood
-        #    of predictions for masked words.
-        # 2. If `lm_labels` is provided we are in a causal scenario where we
-        #    try to predict the next token for each input in the decoder.
         if labels is not None:
             loss_fct = CrossEntropyLoss()  # -100 index = padding token
             masked_lm_loss = loss_fct(prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
             outputs = (masked_lm_loss,) + outputs
 
-        if lm_labels is not None:
-            # we are doing next-token prediction; shift prediction scores and input ids by one
-            prediction_scores = prediction_scores[:, :-1, :].contiguous()
-            lm_labels = lm_labels[:, 1:].contiguous()
-            loss_fct = CrossEntropyLoss()
-            ltr_lm_loss = loss_fct(prediction_scores.view(-1, self.config.vocab_size), lm_labels.view(-1))
-            outputs = (ltr_lm_loss,) + outputs
-
-        return outputs  # (ltr_lm_loss), (masked_lm_loss), prediction_scores, (hidden_states), (attentions)
+        return outputs  # (masked_lm_loss), prediction_scores, (hidden_states), (attentions)
 
     def prepare_inputs_for_generation(self, input_ids, attention_mask=None, **model_kwargs):
```

Comment:
The `prepare_inputs_for_generation` function has to be added to `BertLMHeadModel` so that we can generate with Bert2Bert encoder-decoder. 

We can now split the function into two cases: 1) self.config.is_decoder is False 2) self.config.is_decoder is True and add the function to its corresponding models.

Reply:
It is in `BertLMHeadModel`. Do you mean only the first test should be in `BertLMHeadModel` and the only the second test should be in `BertForMaskedLM`?

Reply:
Sorry my bad! You did add the function. 

Yeah, I think now we can remove some logic from the `prepare_inputs_for_generation` since `config.is_decoder` should not be true anymore for `BertForMaskedLM` and `config.is_decoder` should always be True for `BertLMHeadModel` (since it should use the causal mask` for language generation). But maybe this could be done in a separate PR.

Reply:
Here, this should be enough:

```python 
        effective_batch_size = input_ids.shape[0]

        #  add a dummy token
        assert self.config.pad_token_id is not None, "The PAD token should be defined for generation"
        attention_mask = torch.cat(
            [attention_mask, attention_mask.new_zeros((attention_mask.shape[0], 1))], dim=-1
        )
        dummy_token = torch.full(
            (effective_batch_size, 1), self.config.pad_token_id, dtype=torch.long, device=input_ids.device
        )
        input_ids = torch.cat([input_ids, dummy_token], dim=1)

        return {"input_ids": input_ids, "attention_mask": attention_mask}
```

Reply:
Done.